File: src/server/modules/ServerComponents/ArmorPortal.lua
local Players = game:GetService("Players")
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local ServerStorage = game:GetService("ServerStorage")

local Knit = require(ReplicatedStorage.Packages.Knit)
local Signal = require(Knit.Util.Signal)
local Trove = require(Knit.Util.Trove)
local Component = require(Knit.Util.Component)
local Logger = require(script.Parent.Extensions.Logger)

local IdentityService

local ArmorPortal = Component.new {
	Tag = "ArmorPortal";
	Extensions = {
		Logger,
	};
}

local ARMOR_INFO = require(ServerStorage.ArmorInfo)

function ArmorPortal:Construct()
	self._trove = Trove.new()
	self._debounce = {}

	self.ArmorToSet = self.Instance:GetAttribute("ArmorName")
end

function ArmorPortal:OnTouched(toucher: Part)
	local player = Players:GetPlayerFromCharacter(toucher.Parent)
	if player ~= nil then
		if self._debounce[player.UserId] then return end
		self._debounce[player.UserId] = true

		if player:GetAttribute("Armor") == self.ArmorToSet then
			IdentityService.SetArmor(player, ARMOR_INFO.GetRandomStarter())
		else
			IdentityService.SetArmor(player, self.ArmorToSet)
		end

		task.wait(2)
		self._debounce[player.UserId] = nil
	end
end

function ArmorPortal:Start()
	IdentityService = Knit.GetService("IdentityService")

	self._trove:Connect(self.Instance.Touched, function(...) self:OnTouched(...) end)
end

function ArmorPortal:Stop()
	self._trove:Destroy()
end

return ArmorPortal

File: src/server/components/world/armor-portal.ts
import { BaseComponent, Component } from "@flamework/components";
import { OnStart } from "@flamework/core";
import { Players } from "@rbxts/services";
import { Trove } from "@rbxts/trove";
import { IdentityService } from "server/services/identity-service";

interface Attributes {
	armorToSet: string;
}

@Component({
	tag: "ArmorPortal",
	defaults: {
		armorToSet: "SpiderCloak",
	},
})
export class ArmorPortal
	extends BaseComponent<Attributes, BasePart>
	implements OnStart
{
	private trove = new Trove();
	private debounce = new Array<Player>();

	public constructor(private identityService: IdentityService) {
		super();
	}

	public onStart(): void {
		this.trove.connect(this.instance.Touched, (otherPart) =>
			this.onTouched(otherPart),
		);
	}

	override destroy(): void {
		this.trove.clean();
		super.destroy();
	}

	public onTouched(otherPart: BasePart): void {
		const character = otherPart.Parent as Model | undefined;
		if (!character) return;
		const player = Players.GetPlayerFromCharacter(character);
		if (!player) return;
		if (this.debounce.includes(player)) return;

		const index = this.debounce.push(player) - 1;
		this.identityService.applyArmor(character, this.attributes.armorToSet);

		task.delay(2, () => this.debounce.remove(index));
	}
}
